Don't guard types sometimes

tjhance opened the pull request:
We were doing some unnecessary guards, like guarding the types of constant objects. Fixing this saves about 10 bytes on average per IC, and has a comparable small improvement in performance:

```
       django_template.py             5.1s (6)             5.1s (8)  -1.6%
            pyxl_bench.py             4.2s (6)             4.1s (6)  -2.4%
 sqlalchemy_imperative.py             2.1s (6)             2.0s (6)  -2.5%
        django_migrate.py             2.0s (6)             1.9s (6)  -1.9%
      virtualenv_bench.py             8.4s (6)             8.3s (6)  -0.5%
                  geomean                 3.8s                 3.7s  -1.8%
```

tjhance commented:
Hm so the build is failing, and the culprit appears to be this change:

```
+ if (!cls->is_constant) {
      rewrite_args->obj->addAttrGuard(cls->attrs_offset + HCATTRS_HCLS_OFFSET, (intptr_t)hcls);
+ }
```

The goal was to not guard on the hidden class of builtin types (you can't set attributes on them). I figure if a class is constant, the hidden class shouldn't ever change. I guess this assumption is wrong. Any idea why?

tjhance commented:
oh, do we change the hidden class to be `DICT_BACKED` in the event that someone accesses the `__dict__`? That's a shame :\

tjhance commented:
or does that only happen when someone sets the __dict__? (in which case my previous comment wouldn't apply)

tjhance commented:
OK I think I fixed the problem, I was checking `is_constant` on the wrong thing.

The perf improvement is not as large (but who knows, really, because noise). The IC size improvement is still significant, so I think there will be an even better improvement when we actually get around to making the IC slot sizes smaller.

kmod commented on src/runtime/objmodel.cpp:
```diff
@@ -961,6 +964,7 @@ Box* typeLookup(BoxedClass* cls, llvm::StringRef attr, GetattrRewriteArgs* rewri
                 assert(rewrite_args->obj == obj_saved);
             } else {
                 rewrite_args->obj = rewrite_args->rewriter->loadConst((intptr_t)base, Location::any());
+                rewrite_args->obj_cls_guarded = true;
```
Is this true?  I think you might be allowed to to change the metaclass of a base class...  Maybe this could be `obj_cls_guarded = base->cls->is_constant`?

kmod commented:
I think this is a bit confusing since this PR is based off a commit (adding "-a") that got rewritten and then merged; would you mind rebasing?  I think this applies to some of your other PRs as well (#654, #656, #659).  otherwise lgtm!

tjhance commented:
I rebased and added a check that the base class is a constant class.

Can you double check that all my assumptions look sane and then merge this in?
